Entities/ADoor.cpp

```diff
@@ -1,921 +1,598 @@
+		} else {
+			// Draw the door back if we were indeed temporarily suppressing it before
+			if (m_DoorMaterialDrawn && m_MaterialDrawOverride != enable) {
+				Vector notUsed;
+				m_Door->Draw(g_SceneMan.GetTerrain()->GetMaterialBitmap(), notUsed, g_DrawMaterial, true);
+				// Register that we changed the material layer of the terrain
+				g_SceneMan.GetTerrain()->AddUpdatedMaterialArea(m_Door->GetBoundingBox());
+			}
+			m_MaterialDrawOverride = enable;
+		}
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int ADoor::RemoveAnyRandomWounds(int amount) {
+		float damage = 0;
+		for (int i = 0; i < amount; i++) {
+			// Fill the list of damaged bodyparts
+			std::vector<MOSRotating *> bodyParts;
+			if (GetWoundCount() > 0) { bodyParts.push_back(this); }
+			if (m_Door && m_Door->GetWoundCount()) { bodyParts.push_back(m_Door); }
+
+			// Stop removing wounds if there are not any left
+			if (bodyParts.size() == 0) {
+				break;
+			}
+
+			int partIndex = RangeRand(0, bodyParts.size() - 1);
+			MOSRotating * part = bodyParts[partIndex];
+			damage += part->RemoveWounds(1);
+		}
+		return damage;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void ADoor::GibThis(Vector impactImpulse, float internalBlast, MovableObject *ignoreMO) {
+		if (m_Door && m_Door->IsAttached()) {
+			// Remove the material representation of the door, if any
+			EraseDoorMaterial();
+			m_Door->DeepCheck(true);
+			m_Door->SetPinStrength(0);
+			m_Door->SetVel(m_Vel + m_Door->GetParentOffset() * PosRand());
+			m_Door->SetAngularVel(NormalRand());
+			g_MovableMan.AddParticle(m_Door);
+			RemoveAttachable(m_Door);
+			m_Door = 0;
+		}
+		Actor::GibThis(impactImpulse, internalBlast, ignoreMO);
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void ADoor::UpdateChildMOIDs(vector<MovableObject *> &MOIDIndex, MOID rootMOID, bool makeNewMOID) {
+		if (m_Door) { m_Door->UpdateMOID(MOIDIndex, m_RootMOID, makeNewMOID); }
+		Actor::UpdateChildMOIDs(MOIDIndex, m_RootMOID, makeNewMOID);
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void ADoor::OpenDoor() {
+		if (m_DoorState == STOPPED) {
+			SharedDoorControls();
+			m_DoorState = CLOSING;
+			m_ChangedDirectionAfterStop = (m_DoorState == m_DoorStateOnStop) ? false : true;
+		}
+		if (m_DoorState == CLOSED || m_DoorState == CLOSING) {
+			SharedDoorControls();
+			m_DoorState = OPENING;
+		}
+		m_ResetDefaultTimer.Reset();
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void ADoor::CloseDoor() {
+		if (m_DoorState == STOPPED) {
+			SharedDoorControls();
+			m_DoorState = OPENING;
```

Comment:
I think I understand the rest of the stuff in Open/Close/Stop and SharedDoorControls, but I don't really get why the state is set to OPENING here and CLOSING in the OpenDoor method. Can you explain that to me?

Reply:
It's so the following check is true and door movement can actually happen. It's pretty easy to figure out from looking at the code.

Reply:
lol well apparently I was having trouble. I was pretty tired at that point though, so I'll have to look again and see if I catch on.